### PR TITLE
[Flang][OpenMP] Align  map clause generation and fix issue with non-shared allocations for assumed shape/size descriptor types

### PR DESCRIPTION
**Pull request by agozillon**
This PR aims to unify the map argument generation behavior across both the implicit capture (captured in a target region) and the explicit capture (process map), currently the varPtr field of the MapInfo for the same variable will be different depending on how it's captured. This PR tries to align that across the generations of MapInfoOp in the OpenMP lowering.

Currently, I have opted to utilise the rawInput (input memref to a HLFIR DeclareInfoOp) as opposed to the addr field which includes more information. The side affect of this is that we have to deal with BoxTypes less often, which will result in simpler maps in these cases. The negative side affect of this is that we don't have access to the bounds information through the resulting value, however, I believe the bounds information we require in our case is still appropriately stored in the map bounds, and this seems to be the case from testing so far.

The other fix is for cases where we end up with a BoxType argument into a function (certain assumed shape and sizes cases do this) that has no fir.ref wrapping it. As we need the Box to be a reference type to actually utilise the operation to access the base address stored inside and create the correct mappings we currently generate an intermediate allocation in these cases, and then store into it, and utilise this as the map argument, as opposed to the original.

However, as we were not sharing the same intermediate allocation across all of the maps for a variable, this resulted in errors in certain cases when detatching/attatching the data e.g. via enter and exit. This PR adjusts this for cases

Currently we only maintain tracking of all intermediate allocations for the current function scope, as opposed to module. Primarily as the only case I am aware of that this is required is in cases where we pass certain types of arguments to functions (so I opted to minimize the overhead of the pass for now). It could likely be extended to module scope if required if we find other cases where it's applicable and causing issues.

### CODE REVIEW

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp
Comment (agozillon) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: @jeanPerier @tblah I am not the most familiar with HLFIR (or FIR) yet unfortunately, however, I was interested in knowing if there was any possible side effects of utilising the memref/input from a HLFIR declare operation as opposed to the other possible result which seems to carry more information?

I believe, at least for the moment, we don't need the extra information provided by the HLFIR DeclareOp's other result from my (possibly flawed) understanding, As we already generate and carry the bounds around as part of the MapInfoOp, and we don't need any further information it currently provides from what I can tell. The main benefit of utilising the rawInput (result 1, of the declare op when there, otherwise just the symbol adcress) is it simplifies the maps we generate in certain scenarios as we end up processing less BoxTypes (as in certain cases, the HLFIR declare op's result 0 provides the more complex to map BoxType as opposed to the simpler original input type, presumably to help carry more information).

Essentially wondering if utilising the rawInput in all cases (at least that we currently handle) seems reasonable and sane from a FIR/HLFIR perspective. If we ever required the more complex result, it would be possible to do so, just with a slight extra layer of complexity to the lowering and a (likely mild) performance hit from extra mappings in certain cases.

Comment (tblah) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: Yeah if you are sure you don't need the extra information, using the raw input will avoid unnecessary embox operations etc (for the unused result).

Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: Thank you very much @tblah, at least for the moment I don't believe we need the extra information (and testing against the little local map test suite I have made it seems to function as correctly as it did previously), as we tend to package the bounds as a seperate input to Map Info and at least for the moment that's all we care about, it could change in the future but we can likely cross that bridge if/when we come to a situation that requires it.

Comment (jeanPerier) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: I concur with Tom. If you are not directly mapping this "raw" value to the symbol in the OpenMP context, and are not using hlfir tools with this value, you are fine.

Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: > I concur with Tom. If you are not directly mapping this "raw" value to the symbol in the OpenMP context, and are not using hlfir tools with this value, you are fine.

Thank you very much @jeanPerier! In this case, as the target region is IFA, there's a block argument generation and symbol rebinding that occurs for all map.info operations to corresponding block arguments (that correspond to symbols used inside of the region). However, this binding only extends for the scope of the target region and seems to not pose a problem when utilising the "raw" value as the map input. Would this be something you think would cause an issue?

Comment (jeanPerier) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]: > However, this binding only extends for the scope of the target region and seems to not pose a problem when utilising the "raw" value as the map input. Would this be something you think would cause an issue (from testing so far, it doesn't appear to)?

As long as this new binding inside the OpenMP region happens by generating a new hlfir.declare with the non default lower bounds (also passed as block arguments?), then it is fine, even if the hlfir.declare input is the "raw" value (just like outside of the OpenMP region).

If you tested something like with n different from 1, you are fine:
```
subroutine omp_target_implicit_bounds(n, m)
   integer(8) :: n,m
   integer :: a(n:m)
   !$omp target
      a(11) = 22
   !$omp end target
end subroutine omp_target_implicit_bounds
```

Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:976]:
```
subroutine omp_target_implicit_bounds(n, m)
    integer(8) :: n,m
    integer :: a(n:m)
   !$omp target
       a(4) = 22
       a(15) = 2222
   !$omp end target
    print *, a
! Results in:  22 0 0 0 0 0 0 0 0 0 0 2222
end subroutine omp_target_implicit_bounds

subroutine omp_target_passed_in_implicit_bounds(n, m, a)
    integer(8) :: n,m
    integer :: a(n:m)
   !$omp target
       a(4) = 22
       a(15) = 2222
   !$omp end target
end subroutine omp_target_passed_in_implicit_bounds

program test
    integer(8) :: n,m
    integer :: array(20)
    n = 4
    m = 15
    call omp_target_implicit_bounds(n,m)
    call omp_target_passed_in_implicit_bounds(n, m, array)
    print *, array
    ! Results in: 22 0 0 0 0 0 0 0 0 0 0 2222 0 0 0 0 0 0 0 0
end program
```

Thank you very much @jeanPerier, I took your test and expanded it a bit into a full program, it seems to function as I would expect (although, I am far from a Fortran expert) providing the same results when there is a target region and when there's not.

From IR introspection, it does appear we create a hlfir.declare inside of the target region, and specify a shapeshift (lower bound and extent) calculated from mapped in bounds information.

So, it seems if the above assessment is correct, using the raw input for the moment is fine. At least from a correctness perspective, I am not sure which would come with a higher performance tax (if either).

**Comment by agozillon** [comment]
Small ping on this for some reviewer attention please if at all possible! Thank you very much ahead of time.

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:86]:
```suggestion
      // If we have already created a local allocation for this BoxType,
```

File: flang/test/Lower/OpenMP/target.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/target.f90:462]: How is the removal of the `stride_in_bytes` related to this patch?

**Review by skatrak — COMMENTED** [review]
Thank you Andrew, only minor comments from me. There are some tests that you had to update, but are there any new cases handled this patch for which adding a new test makes sense?

**Review by agozillon — COMMENTED** [review]
No new cases really being handled, just an overall simplification of the code and making both implicit map and explicit maps aligned. However, there's also a minor bug fix via sharing locally made temporary allocations that are load/stored into (fixed by both the OMPMapInfoFinalization pass changes and to a lesser extent by avoiding some cases with the swap to raw input, but the former should hopefully catch any cases the latter doesn't avoid). I am more than happy to add a test case for the bug fix to the runtime tests as that's primarily where the problem will be detected, as it occurs as a runtime mapping error!

I'll update the test and have a look into swapping back to a module only pass, it'll require a bit more complexity in the pass as I'll have to work out a way to appropriately track which allocations belong to which function. Likely not too difficult though!

File: flang/include/flang/Optimizer/Transforms/Passes.td
Comment (agozillon) [reply flang/include/flang/Optimizer/Transforms/Passes.td:344]: Actually looking a bit further into it, it's still ran as a FuncOp pass, I've just made it a little bit more explicit with these additions. Not what I would have expected in this case! But in either case, happy to make it a ModuleOp pass, I've yet to encounter any race condition issues, but considering the pass is in flux quite a bit, it's likely better safe than sorry and we can convert it to a FuncOp pass and iron out any possible issues when it's a little more stable.

**Comment by agozillon** [comment]
Small ping for some reviewer attention on this PR please, if at all possible. Thank you ahead of time as always! :-)

File: flang/include/flang/Optimizer/Transforms/Passes.td
Comment (skatrak) [reply flang/include/flang/Optimizer/Transforms/Passes.td]: I don't have an issue with it running in parallel for each function. If it already used to run like that, then my guess is that it's not a problem here, so it's fine by me either way. The main thing to be wary of in that case is accessing other functions or global operations, and perhaps creating new global symbols or making changes to module attributes.

**Review by skatrak — APPROVED** [review]
Thank you Andrew, this LGTM. I'd suggest waiting for at least a second reviewer on this one, because it's likely some details here are flying over my head... 😅

**Comment by agozillon** [comment]
> Thank you Andrew, this LGTM. I'd suggest waiting for at least a second reviewer on this one, because it's likely some details here are flying over my head... 😅

No worries, I'll do that! Thank you very much for the review, it's greatly appreciated!

If one other person could volunteer and give a review of this PR it would be greatly appreciated, it would be nice to land it in the near future :-)

**Review by jsjodin — APPROVED** [review]
LGTM

**Review by TIFitis — APPROVED** [review]
LGTM :)

**Comment by agozillon** [comment]
Thank you very much everyone for the reviews! I'll update and land this tomorrow or sometime during this week.

**Comment by agozillon** [comment]
Updated to resolve conflicts, will now wait until it passes at least the Linux CI before landing the PR, thank you for the reviews everyone it's greatly appreciated.

**Pull request merged**
